include/zephyr/linker/common-rom/common-rom-bt.ld

```diff
@@ -1,6 +1,8 @@
 /* SPDX-License-Identifier: Apache-2.0 */
 
 	ITERABLE_SECTION_ROM(bt_l2cap_fixed_chan, 4)
+	ITERABLE_SECTION_ROM(l2cap_userdata_destroy, 4)
```

Comment:
I would prefer we add a `destructor` field to `bt_conn_tx` and `l2cap_tx_meta_data`. Is there a reason we should go for this iterable section approach?

Reply:
This could be an option. I used the iterable sections to register the cleanup callbacks at compile time, but dont have a good reason. Your suggestion may be cleaner.

Reply:
An issue here may be that we need to pass a cleanup function to the `_cb` functions in conn and l2cap. Other than that it would be nice to simply do `cleanup(user_data)` in the failure cases and `cb(user_data)` in the success cases.

Reply:
@jhedberg @Thalley and @jori-nordic can you weigh in on this choice between the two implementations?

Reply:
I think adding a `destructor` field would be easier to understand.

Reply:
Assuming that all contexts will share the same destructor, I am OK with either approach. 

If we want to use different destructors, then the field approach makes more sense IMO (but that may be a double-edged sword too).

Reply:
Observation: The current approach relies on the destructors recognizing their applicability by examining the pointer value (e.g. by using ARRAY_OF). This means that two destructors may not use the same memory pool, otherwise they will incorrectly activate. It therefore is not possible to use, as user_data, the heap, global net-buf pool, gobal-ish application defined pools, etc.

Reply:
> An issue here may be that we need to pass a cleanup function to the `_cb` functions in conn and l2cap. Other than that it would be nice to simply do `cleanup(user_data)` in the failure cases and `cb(user_data)` in the success cases.

This is an interesting discussion. I can think of two mitigations for this:
 - Using the callback as the destructor as well. `cb(FAILURE, user_data)` and `cb(SUCCESS, user_data)`. Does not need to be only FAILURE/SUCCESS of course.
 - Adding a Destructable interface to the object-to-be-destroyed. See the sketch below. It's a bit verbose. But it's a nice interface.
   ```c
   /* Wrapper for a destructor function pointer used to avoid a potential mess with auto-dereferencing function pointers. */
   struct Destructible {
       void (*destructor)(void * obj);
   }
   /* Structs implementing this interface must have the Destructible in a field named `interface_destructible`. */
   #define downcast_destructible(obj) (&(obj).interface_destructible)
   #define upcast_destructible(obj_type, destructible) CONTAINER_OF(destructible, obj_type, interface_destructible)
   
   
   struct app_user_data_t {
       const struct Destructible interface_destructible;
       ...
   }
   /* Implements destructible for struct my_user_data. */
   void app_user_data_t_destructor(struct Destructible * destructible) {
       app_user_data_t * user_data = upcast_destructible(app_user_data_t, destructible);
       ...
   }
   
   
   void cb(struct Destructible * destructible) {
       app_user_data_t * user_data = upcast_destructible(app_user_data_t, destructible);
       ...
   }
   struct app_user_data_t cb_user_data = {
       .interface_destructible = { app_user_data_t_destructor };
   };
   void app_code() {
       l2cap_code(..., cb, downcast_destructible(&cb_user_data))
   }
   ```

Reply:
I'd prefer if we can avoid the linker section solution. That primarily makes sense when there's a fair chance that the data you pass to the callbacks may be useful for most/all of the callbacks, however here you're forced to iterate to find that one single callback that's the correct one. So I think adding another callback to the existing struct makes more sense.